### PR TITLE
Add support for hardware and board initialisation overrides.

### PR DESCRIPTION
**Pull request by tzarc**
## Description

ChibiOS board files aren't normally able to be overridden without making duplicates of the definitions, and modding them.

This PR introduces some new API methods and provides a way to override the initialisation files such that hooking the early bootup sequence can be achieved whilst retaining the original ChibiOS board definitions.

Effective changes, replacing modifications for each board definition's `board.c`:
 - `__early_init` has been overridden, and now has qmk api equivalents `early_hardware_init_pre` and `early_hardware_init_post`
 - `boardInit` has been overridden, and now has qmk api equivalent `board_init`

See documentation for notes on the constraints for usage -- any board developer that has previously overridden these will likely know the constraints already.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by tzarc** [comment]
Any ideas where these new APIs should be documented?

**Comment by tzarc** [comment]
I've made the `early_hardware_init_pre` a no-op for now, so that we're not attempting to enter bootloader twice. Will have a followup set of PRs which migrate existing usage of the `__early_init` currently in use by keyboards, and once all the migration is complete, it can probably be swapped back in.

**Comment by tzarc** [comment]
Initial testing makes it possible to remove the custom board definitions from the test projects I've been working with (`BOARD = ......` in rules.mk), dropping back to the ChibiOS-supplied ones.

**Comment by tzarc** [comment]
Seems to work fine for both Proton-C (F303), custom L082, custom G431, custom G474.

File: tmk_core/protocol/chibios/main.c [hunk 1]
```diff
@@ -101,6 +101,39 @@ void midi_ep_task(void);
 //   }
 // }
 
+/* Early initialisation
+ */
+__attribute__((weak)) void early_hardware_init_pre(void) {
+#if FALSE && defined(STM32_BOOTLOADER_ADDRESS) // remove the FALSE when we've migrated all existing __early_init invocations
```
Comment (zvecr) [inline tmk_core/protocol/chibios/main.c:107]: Is there a better option that FALSE here so someone can prep for the eventual refactor? (also name are difficult...)
```suggestion
#if defined(ENABLE_STM32_BOOTLOADER_ADDRESS) && defined(STM32_BOOTLOADER_ADDRESS) // remove the FALSE when we've migrated all existing __early_init invocations
```

Comment (tzarc) [reply tmk_core/protocol/chibios/main.c:107]: I'll end up doing the refactoring at some point after this regardless, specifically for the dual-bank flash version that requires the extra GPIO. Will leave this name for now and potentially refactor as required at the time.

**Pull request merged**